internal/provider/resource_vault_cluster.go

```diff
@@ -680,6 +723,48 @@ func resourceVaultClusterUpdate(ctx context.Context, d *schema.ResourceData, met
 		return diag.Errorf("unable to retrieve Vault cluster (%s): %v", clusterID, err)
 	}
 
+	// on update, delete plugins that were removed from plugin config. Add all plugins in new config
+	if d.HasChange("vault_plugin") {
+		old, _ := d.GetChange("vault_plugin")
+
+		oldPlugins := old.([]interface{})
+
+		for _, oldP := range oldPlugins {
+			config, ok := oldP.(map[string]interface{})
+			if !ok {
+				return diag.Errorf("could not parse old plugin config: %v", err)
+			}
+
+			pluginName := config["plugin_name"].(string)
+			pluginType := config["plugin_type"].(string)
+
+			// if plugin in old config is not found in the new config, delete the plugin
+			found := false
+			for _, plugin := range newPluginConfig {
+				if strings.EqualFold(pluginName, plugin.PluginName) && strings.EqualFold(pluginType, plugin.PluginType) {
+					found = true
+				}
+			}
+
+			if !found {
+				req := &vaultmodels.HashicorpCloudVault20201125DeletePluginRequest{PluginName: pluginName, PluginType: pluginType}
+				_, err := clients.DeletePlugin(ctx, client, clusterLocationShared, clusterID, req)
+				if err != nil {
+					return diag.Errorf("error deleting plugin (%s) on Vault cluster (%s): %v", pluginName, clusterID, err)
+				}
+			}
+
+		}
+
+		// add all plugins in new plugin config
+		for _, plugin := range newPluginConfig {
+			_, err := clients.AddPlugin(ctx, client, clusterLocationShared, clusterID, plugin)
+			if err != nil {
+				return diag.Errorf("error adding plugin (%s) to Vault cluster (%s): %v", plugin.PluginName, clusterID, err)
+			}
+		}
+	}
+
 	if err := setVaultClusterResourceData(d, cluster); err != nil {
```

Comment:
Do you need to update the `setVaultClusterResourceData` function to set the vault plugin info there as well? https://github.com/hashicorp/terraform-provider-hcp/blob/678ec3bd0dad3914655b6a0881f68a9c9ae88158/internal/provider/resource_vault_cluster.go#L810

I was planning to do this for the proxy changes since we appear to set all the fields there (calling the above function) at the end of create, update, and read.

Reply:
i honestly didn't pay too much attention to that function and just left it as the last call. I can move it before the plugin stuff, if that makes it more clear

Reply:
Oh I think it's ok to leave it as the last call, but I was wondering if you need to update the contents of the function itself? Like inside the function, we set each individual key to value of `cluster.Field`, example: https://github.com/hashicorp/terraform-provider-hcp/blob/678ec3bd0dad3914655b6a0881f68a9c9ae88158/internal/provider/resource_vault_cluster.go#L848-L867

Reply:
Did a little more digging. I think in order to set a new field in `setVaultClusterResourceData`, you'd also need to specify the new field in `data_source_vault_cluster` - [this file](https://github.com/hashicorp/terraform-provider-hcp/blob/main/internal/provider/data_source_vault_cluster.go#L18) which corresponds to the data source specification [here](https://registry.terraform.io/providers/hashicorp/hcp/latest/docs/data-sources/vault_cluster). 

In my own [branch](https://github.com/hashicorp/terraform-provider-hcp/compare/hvt-4182-http-proxy-for-hcpv-in-hcp-tfp?expand=1) for proxy changes (feel free to check it out but it's def still a WIP), I originally tried adding my fields in `setVaultClusterResourceData` without adding them to the data source, which failed. After adding to data source, setting the new fields in `setVaultClusterResourceData` worked.

Reply:
good callout, I can try to add it

Reply:
updated